assets/lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver/templates/kube-apiserver.yaml

```diff
@@ -81,6 +109,67 @@ spec:
         - name: ssl-certs-host
           mountPath: /etc/ssl/certs
           readOnly: true
+        - name: data
+          mountPath: /run/kube-apiserver
+      - name: haproxy
+        image: haproxy:2.1.3-alpine
+        volumeMounts:
+        - name: data
+          mountPath: /run/kube-apiserver
+        command:
+        - /bin/sh
+        - -c
+        - |
+          set -x && \
+          export ADDRESS=$(cat /run/kube-apiserver/address) && \
+          if [ -z $ADDRESS ]; then \
+            echo "ADDRESS not found" && \
+            exit 1; \
+          fi && \
+          echo "defaults \
+            # Do TLS passthrough \
+            mode tcp \
+            # Required values for both frontend and backend
+            timeout connect 5000ms
+            timeout client 50000ms
+            timeout server 50000ms
+
+          frontend kube-apiserver-internal
+            bind $POD_IP:{{ template "port" . }}
+            default_backend kube-apiserver
+
+          {{- if .Values.apiserver.exposeOnAllPorts }}
+          frontend kube-apiserver-external
+            bind 0.0.0.0:6443
+            default_backend kube-apiserver
+
+          {{- end }}
+          backend kube-apiserver
+            server 1 $ADDRESS:{{ template "port" . }} check" > /run/kube-apiserver/haproxy.cfg && \
+          echo "Connecting to $ADDRESS:{{ template "port" . }}" && \
+          until nc -zv $ADDRESS {{ template "port" . }}; do sleep 1; done && \
```

Comment:
This is to make sure, that `kube-apiserver` container is up and running before we start HAProxy process, as we use TCP readiness probe and it will be "UP" as long HAProxy process starts, as HAProxy does not reject incoming connections, if no backend is available, but returns 503. Also yes, if for whatever reason that happens, then also pod checkpointer will kick in, and as HAProxy does not bind then yet, it will start just fine.

Reply:
Okay, so this logic is to not show as ready the pod when the api server isn't ready?

If the api server crashes, this will still happen (as HAProxy will still be running), right?

Reply:
I mean, that can only be "meaningful" in the initialization, as if the API server crashes the HAProxy will still be running and, if IIUC the comment: https://github.com/kinvolk/lokomotive/pull/32/files#diff-12be5db389a0fefe0038d10211a890b4R196-R199

HAProxy will not reject connections to API server even if it's down.

Is there some value on having this? IIUC, no readinessProbe is set in the non-HA case, and this infinite loop can only make sure it connected once, but what if it connects and API server crashes just immediately after?

I'm trying to see why this is useful in practice. Has it been useful to you when developing this?

Reply:
>Is there some value on having this? IIUC, no readinessProbe is set in the non-HA case, and this infinite loop can only make sure it connected once, but what if it connects and API server crashes just immediately after?

Then it will be restarted by the `kubelet`. But in general, `kube-apiserver` won't start accepting connections until it makes sure, that all flags etc are correct.

>I'm trying to see why this is useful in practice. Has it been useful to you when developing this?

We should have readiness probes, as they are used by helm to determine if the upgrade process succeeded. Without the readiness probe, upgrade is always assumed as successful, see #83 for more details.

Reply:
Oh, that is a very good reason! Can you please add a comment above the readinessProbe? I can totally see my future self removing that (as it seems pointless as it only tests HAProxy). I don't want to remove that by mistake :-D

Reply:
Added:
```
          # We use TCP readiness probe and HAProxy does not reject connections if no backend is available,
          # so we wait until kube-apiserver is available here, so readiness of haproxy container represents
          # readiness of kube-apiserver, as kube-apiserver cannot have readiness probe set, as it listens
          # on random IP address.
```

Reply:
It's missing the most important part:
```Readiness probe here tests that at least the apiserver was up at least once (because HAProxy waiting for apiserver to start), after that it will mean just that HAProxy is up. But this is needed for helm to perform atomic upgrades, helm can't do atomic upgrades in components without readinessProbe
```
?

Reply:
Added to readinessProbe comment.